### DISCUSSION TITLE
Login with FedCM or IndieAuth

### DISCUSSION THREAD

**Discussion by erlend-sh** [bluesky-social/atproto#2796] [General Discussion]
https://blog.erlend.sh/indie-social-sign-in-could-go-mainstream

> [Domain names as handles](https://bsky.social/about/blog/4-28-2023-domain-handle-tutorial) is a flagship feature of the Bluesky network. It follows rather naturally that users ought also be able to log into the network using their own domains.

https://bsky.app/profile/erlend.sh/post/3l3itw5fqil2f

**Comment by bnewbold** [comment]
This looks intriguing! In the context of atproto handles, it seems like this could work at least two ways: folks could log in to their atproto PDS using their domain using FedCM/IndieAuth (aka, atproto PDS as the Relying Party); or they could use an atproto PDS (possibly self-hosted, possibly on their own domain) to log in to other services using FedCM/IndieAuth. Does that sound right?

We have made a big push/investment in getting out the atproto profile of OAuth (based on OAuth 2.1 and client metadata), which is related. Have you checked that out?

* https://github.com/bluesky-social/atproto-website/pull/326
* https://docs.bsky.app/docs/advanced-guides/oauth-client

It ends up achieving some of the same goals, and allows both authz and authn use-cases, but we didn't end up with direct OIDC compatibility.

**Reply by erlend-sh** [reply]
> it seems like this could work at least two ways:
> 1. folks could log in to their atproto PDS using their domain using FedCM/IndieAuth (aka, atproto PDS as the Relying Party);
> 2. or they could use an atproto PDS (possibly self-hosted, possibly on their own domain) to log in to other services using FedCM/IndieAuth.
> 
> Does that sound right?

Yep! That’s an important clarification that I didn’t delve into in my post, cheers.

I am indeed following the atproto OAuth work, but only as a product designer. For the engineer’s view I defer to my internet-colleague @anderspitman who’s already made tangible progress with a working proof-of-concept: https://github.com/w3c-fedid/idp-registration/issues/2#issuecomment-2336490996

**Reply by aaronpk** [reply]
Hey @bnewbold! I wrote up this guide for how to layer FedCM and IndieAuth. https://indieweb.org/FedCM_for_IndieAuth IndieAuth has a lot of similarities to what you've got going with OAuth now. Take a look at the guide since most of it will apply to your implementation too. There's no requirement to use OpenID Connect in FedCM, it's meant to be usable with any protocol, and I've been working on writing up guides for OAuth, IndieAuth, and OpenID Connect.

**Reply by anderspitman** [reply]
@bnewbold sorry for the delay. I just read through https://github.com/bluesky-social/atproto-website/blob/bnewbold/spec-oauth/content/specs/auth.md. Excellent document. Honestly it's a pretty good summary of the state of OAuth2 and how to design a modern OAuth2 profile.

With FedCM, what we're really hurting for currently is RP implementations. [Google's position](https://github.com/w3c-fedid/idp-registration/issues/2#issuecomment-2334799426) is essentially that IdP's and users are likely to embrace FedCM, but there's not enough evidence that RPs will also do so.

So we're looking for prominent RPs to jump on board. It's a chicken/egg problem, and no guarantee the protocol gets adopted in browsers, but I personally believe it would be a big win for the open web if it did, and it's worth sacrificing to try and make it happen. So I've invested a significant amount of time in it so far.

Would be happy to talk more about the potential for atproto to get involved.

I have some additional general questions, but in the interest of keeping this focused on FedCM I posted them here: https://github.com/bluesky-social/atproto/discussions/2656#discussioncomment-10642015

**Reply by bnewbold** [reply]
@anderspitman we are pretty interested in FedCM generally, but still trying to clarify what would be more helpful from us: are you hoping that atproto PDS instances would be relying parties, or that folks building atproto apps would be relying parties? Or both/either?

We don't support external IdPs of any kind yet in our PDS implementation, which is both what we run for the millions of users on our servers, and most of the other instances in the atproto network. We think it is fine and great for folks to use IdPs for PDS account signup and login, but it will be a chunky piece of work for us to land and isn't on our near-term roadmap. On the other hand, having a pattern for app developers (such as https://smokesignal.events/, https://whtwnd.com/, https://frontpage.fyi/) to support FedCM as a smoother login flow would be great. And https://bsky.app might be in that category. But to clarify, all of these atproto services require a PDS account somewhere; it doesn't make sense at the protocol layer to log in to bsky.app with IndieAuth from your own domain, you need a PDS account in the mix.

**Reply by aaronpk** [reply]
@bnewbold what we're missing as far as the FedCM spec development is concerned is demonstration of interest from relying parties for this feature, so that could absolutely be atproto apps as relying parties.

**Reply by anderspitman** [reply]
I want to make sure I'm understanding correctly. When I log in to Bluesky, I see an option to select my hosting provider (which is the PDS, right?), with Bluesky Social as the default. I'm assuming if I selected my own server, I would first be redirected in some way and have to log in to it, then be redirected back and also provide my Bluesky credentials for the app? Or would the app piggyback off my PDS session and log me in automatically? Sorry, I haven't made time to dig into the ATProto architecture enough yet.

Basically I'm wondering if identity is anchored in the PDS, maybe it makes the most sense for the PDS to be a FedCM IdP, and for ATProto apps to be FedCM RPs. This feels fairly intuitive to me, where your identity is closely tied to where your data lives.

**Reply by bnewbold** [reply]
Just want to say that while I haven't had time to follow-up on this, we are very very interested in FedCM. Please do poke again or reach out to set up a meeting if there is timeline urgency.

**Reply by samuelgoto** [reply]
I'm late to this party, but wanted to help in whichever way I can here! I'm a big blue sky fan (https://bsky.app/profile/sgo.to) and would love to find ways to help whichever way I can as a browser engineer!!!

Aside from facilitating login (which, in and on itself is a big deal, I think), here is another idea that I'm excited about for Blue Sky:

https://github.com/w3c-fedid/idp-registration/issues/15

> Basically I'm wondering if identity is anchored in the PDS, maybe it makes the most sense for the PDS to be a FedCM IdP, and for ATProto apps to be FedCM RPs. This feels fairly intuitive to me, where your identity is closely tied to where your data lives.

This makes sense to me! Does anyone know what would be missing here to accomplish this?

> Please do poke again or reach out to set up a meeting if there is timeline urgency.

Just checking, but would you accept PRs?

**Reply by bmann** [reply]
Forking the Bluesky maintained PDS and making a proof of concept would be the path forward. It’s pretty easy to run a PDS.

**Comment by erlend-sh** [comment]
We’ve got some more movement on this @samuelgoto (also your bsky handle is still invalid and unmentionable!)

https://bsky.app/profile/tom.sherman.is/post/3lpfc73htyk2a

**Comment by erlend-sh** [comment]
This gets us one step closer:

- https://github.com/bluesky-social/atproto/pull/4060

**Comment by erlend-sh** [comment]
There’s now a [handful of new PDS implementations out there](https://bsky.app/profile/erlend.sh/post/3mc7rx5mtqk2r) that are minimally operational already. Seeing as these projects control the whole stack they can move a lot faster than any upstream contribution of FedCM to atproto-official ever could. 

Big opportunity here for community devs to pioneer a breakthrough feature for atproto.

Bryan also expressed official interest in [FedCM for 2026](https://bsky.app/profile/bnewbold.net/post/3m7hm5mugys2y). Ditto [from the ActivityPub camp](https://hedgedoc.socialweb.coop/MoKSsjLPRwyDoDpKodH_mQ#FedCM-Discussion).

For widespread use, I believe the main blocker is the inability to unilaterally [allow external IdPs](https://github.com/w3c-fedid/idp-registration/issues/2), but that doesn’t block the development of a proof-of-concept.

**Reply by bnewbold** [reply]
To confirm, Bluesky is quite interested and we are putting this on our protocol roadmap for early 2026.

**Reply by wdantuma** [reply]
A first step in supporting FedCM is accepting tokens signed with the users signing key as a means of authentication at the PDS / entryway.

I suggest using the ```id_token_hint``` in the OAuth authentication flow for this, support is already partly implemented in the reference PDS.

The IdToken used in the ```id_token_hint``` as used in the "ATproto OAuth profile" would not be an IdToken as specified in OIDC but a JWT token to be only used in the OAuth code flow bound to the client (```client_id```) and /authorize endpoint (```aud```)

Below a proposed minimal list of claim's for such a token

```json
{
  "iss": "did:plc:r6qvtfm3fqlz3vgo3q4uvoe4",
  "sub": "did:plc:r6qvtfm3fqlz3vgo3q4uvoe4",
  "aud": "https://as.example.com/oauth/authorize",
  "exp": 1769351164,
  "iat": 1769351134,
  "scope": "atproto",
  "client_id": "https://rp.example.com/oauth-client-metadata.json"
}
```
The ```iss``` and ```sub``` claims are equal because in ATproto the identifier for the public keys and the user's identifier are both the user's DID.

We could also choose to allow the token the be signed with one of the user's rotation keys, this opens up the possibility to use an external app (Atmosphere wallet ? ) used for maintaining the user's rotation keys to issue the idToken.

PR #4575 add's support for this method to the reference PDS and a demonstration what it looks like used in a FedCM flow can be seen [here](https://bsky.app/profile/willem.dobs.nl/post/3mcr73bs3pk2t)

Next step would be implementing the FedCM config and assertion endpoints in the PDS

**Comment by jimray** [comment]
We're happy to announce that we're sponsoring @ThisIsMissEm, who will be working on FedCM across the open social web, including atproto. Marking this discussion as resolved for now, there will be surely more to come as work progresses on FedCM.

Blog post announcing Bluesky's partnership: https://atproto.com/blog/working-to-decentralize-fedcm

Follow Emelia's work: https://decentralizing-fedcm.leaflet.pub/

**Reply by erlend-sh** [reply]
Huge congrats and thanks to Emelia for driving this work onwards, and major kudos to the Bluesky folks for making this a protocol-agnostic initiative. This literally couldn’t have worked out any better!

**Reply by samuelgoto** [reply]
Sweet!!!

**Reply by anderspitman** [reply]
Congrats @ThisIsMissEm! Really excited to hear this.
